tokio/src/sync/semaphore.rs

```diff
@@ -123,6 +123,79 @@ use std::sync::Arc;
 /// # }
 /// ```
 ///
+/// ## Prevent tests from running in parallel
+///
+/// By default, Rust parallelly runs tests in the same file. However, In some cases, running two tests in parallel may lead to some problems.
+/// For example, this can happen when tests use the same database.
+///
+/// Consider the following scenario:
+/// 1. `test_insert`: Inserts a key-value pair into the database, then retrieves the value using the same key to verify the insertion.
+/// 2. `test_update`: After insertion, this test updates the key's value to a new one and verifies if the value has been accurately updated.
+/// 3. `test_others`: Other tests that dosen't modify the database state, thus can run parallely with previous tests.
+///
+/// For this example, `test_insert` and `test_update` need to run in sequence to ensure they can consistently pass., but it's doesn't matter which one runs first.
+/// Leveraging a semaphore with a single permit elegantly addresses this challenge.
+///
+/// ```
+/// #[cfg(test)]
+/// mod tests {
+///    use tokio::sync::Semaphore;
```

Comment:
It seems like this `#[cfg(test)]` is preventing the test from being compiled in CI. I think it would be better to remove the `mod tests` and just show its contents instead.

This will probably cause compilation errors because the database types don't exist. However, you can add some lines starting with `#` that defines a hidden database that's just based on a hash map. Lines starting with `#` do not show up in the generated documentation.

Reply:
If you run into problems with getting this to work, feel free to ask questions here or message me on our Discord server. You can use the `cargo test --doc --all-features` command to compile the test locally.

Reply:
Seems like that still wasn't enough. Can you try the hack described [here](https://users.rust-lang.org/t/tests-in-doctests/100614)?

Reply:
Sorry, you mean this example should pass the doctest?

Reply:
Yeah, all examples should pass doctests unless there's a really good reason otherwise.

Reply:
Now it can pass doctest :)